plugins/ingestion-kinesis/src/main/java/org/opensearch/plugin/kinesis/KinesisShardConsumer.java

```diff
@@ -0,0 +1,243 @@
+
+        KinesisClientBuilder kinesisClientBuilder = KinesisClient.builder()
+            .region(Region.of(config.getRegion()))
+            // TODO: better security config
+            .credentialsProvider(StaticCredentialsProvider.create(create(config.getAccessKey(), config.getSecretKey())));
+
+        if (config.getEndpointOverride() != null && !config.getEndpointOverride().isEmpty()) {
+            try {
+                kinesisClientBuilder = kinesisClientBuilder.endpointOverride(new URI(config.getEndpointOverride()));
+            } catch (URISyntaxException e) {
+                throw new RuntimeException("Invalid endpoint override: " + config.getEndpointOverride(), e);
+            }
+        }
+
+        return kinesisClientBuilder.build();
+    }
+
+    @Override
+    public List<ReadResult<SequenceNumber, KinesisMessage>> readNext(
+        SequenceNumber sequenceNumber,
+        boolean includeStart,
+        long maxMessages,
+        int timeoutMillis
+    ) throws TimeoutException {
+        List<ReadResult<SequenceNumber, KinesisMessage>> records = fetch(
+            sequenceNumber.getSequenceNumber(),
+            includeStart,
+            maxMessages,
+            timeoutMillis
+        );
+        return records;
+    }
+
+    @Override
+    public IngestionShardPointer earliestPointer() {
+        return getSequenceNumber(ShardIteratorType.TRIM_HORIZON, null, 0);
+    }
+
+    @Override
+    public IngestionShardPointer latestPointer() {
+        return getSequenceNumber(ShardIteratorType.LATEST, null, 0);
+    }
+
+    private List<Record> fetchRecords(ShardIteratorType shardIteratorType, String startingSequenceNumber, long timestampMillis, int limit) {
+        // Get a shard iterator AFTER the given sequence number
+        GetShardIteratorRequest.Builder builder = GetShardIteratorRequest.builder()
+            .streamName(config.getStream())
+            .shardId(kinesisShardId)
+            .shardIteratorType(shardIteratorType);
+
+        if (startingSequenceNumber != null) {
+            builder = builder.startingSequenceNumber(startingSequenceNumber);
+        }
+
+        if (timestampMillis != 0) {
+            builder = builder.timestamp(Instant.ofEpochMilli(timestampMillis));
+        }
+
+        GetShardIteratorRequest shardIteratorRequest = builder.build();
+
+        GetShardIteratorResponse shardIteratorResponse = kinesisClient.getShardIterator(shardIteratorRequest);
+        String shardIterator = shardIteratorResponse.shardIterator();
+
+        if (shardIterator == null) {
+            return new ArrayList<>();
+        }
+
+        // Fetch the next records
+        GetRecordsRequest recordsRequest = GetRecordsRequest.builder().shardIterator(shardIterator).limit(limit).build();
+
+        GetRecordsResponse recordsResponse = kinesisClient.getRecords(recordsRequest);
+        List<Record> records = recordsResponse.records();
+        return records;
+    }
```

Comment:
A Kinesis consumer, in steady state, should keep track of the next shard iterator returned in `GetRecordsResponse` and should use that to fetch the next batch.

In general, we should avoid calling `GetShardIterator` unless we need to reposition the poller (either on startup/recovery, or because we've explicitly moved the pointer).

Without changing the signature of `readNext`, maybe we can remember the last `SequenceNumber`, and if we're being asked to read after that, we can reuse the last returned `ShardIterator`?

Reply:
I see. Good to know this.

Reply:
one challenge with the existing design is that the flow control is managed by the stream poller. As a result, the `readNext` takes `SequenceNumber sequenceNumber` as input, and I don't think it's possible to relate this `sequenceNumber` with the returned `ShardIterator `.

one way to address this is to add another method to the consumer interface of `readNext` without the starting pointer, which implicitly reads from the previous offset. what do you think?

Reply:
Updated this. @msfroh PTAL

Reply:
Yeah -- I think the `readNext` without a sequence number is cleaner. Thanks!